### ISSUE TITLE
Notion Trigger (Beta) triggers doesn't trigger sometimes

### ISSUE THREAD

**Issue by tonidas** [n8n-io/n8n#6604]
**Describe the bug**
It seems that the Notion Trigger doesn't get triggered in some specific time-based scenarios.

**To Reproduce**

1. Create a Notion Trigger for Page Updated in Database
2. Update a select field of a Page of the related Database
3. See that nothing happens, nothing gets triggered. If something happens. wait a minute and go back to step one

**Expected behavior**
That the trigger gets triggered when I change something in a Notion Page

**Environment (please complete the following information):**

- OS: Ubuntu Linux 22.04
- n8n Version 0.222.3
- Node.js Version v16.19.1
- Database system Postgres
- Operation mode   EXECUTIONS_MODE: queue,  EXECUTIONS_PROCESS: own

**Additional context**
I've looked at the source code and it looks to be related to the "last minute" query issue at Notion API. (https://github.com/n8n-io/n8n/blob/master/packages/nodes-base/nodes/Notion/NotionTrigger.node.ts)
I've looked at the newer release's change log and issues related to Notion but couldn't find something similar.

**Comment by tonidas** [comment]
Three possibilities came to my mind:
1. The Notion API request (querying the latest stuff that has been changed since the last minute) is made just when the Page change occurs, somehow leading to a racing condition
2. Differences between the Notion clock and the N8N server clock that could lead to a minute of empty events
3. A bug on the N8N that leads to a good event being discarded

How should I debug the trigger pooling code? I can't see a line related to the pooling on the worker's logs, even with active debug logging.

**Comment by tonidas** [comment]
I was able to reproduce it in a controlled way:

1. Create a Notion Trigger for "Page Updated in Database"
2. Update a select field of a Page (let's call it Page "A") of the related Database.
3. Wait until the Workflow gets triggered
4. Wait a couple of minutes
5. Update a select field on the Page A again
6. Wait endlessly, as this event will be discarded!

Doing some debug, the problem is definitely here: https://github.com/n8n-io/n8n/blob/master/packages/nodes-base/nodes/Notion/NotionTrigger.node.ts

The `possibleDuplicates`  variable is not cleaned up and if you update the same page after some minutes, the code will exclude the entry, thinking that it's a duplicate.

This `possibleDuplicates` approach was introduced on this PR: https://github.com/n8n-io/n8n/pull/2817

Apart from the fact that `possibleDuplicates`, I think that there is another issue. This PR makes the pooling ignore changes made in the same minute on the same page.

Lets say that I have a Page A, and this sequence of events:
1. Event Alpha, 00:01:15, I edit Page A and set its attribute to Xalala
2. Event Gama, 00:01:30, polling code runs and get the Page A change (Event Alpha), stores Page A id into `possibleDuplicates`
3. Event Beta, 00:01:45, I edit Page A and set its attribute to Xelele
4. Event Zeta, 00:02:30, polling code runs and get the Page A change (Event Beta), and ignores it, as the minute of the events were the same: 00:01

TL;DR I definitely found a bug: `possibleDuplicates` is not being cleaned after the next execution without updates. But even if that's fixed I believe that there is another bug regarding the `possibleDuplicates` logic: it's ignoring two events on the same minute.

@knshiro Dear contributor, first I would like to thank you for all your contributions to the community! Secondly, as the mentioned PR was yours, can you kindly help me out on this? If you validate my theories, I could definitely make a PR!

**Comment by knshiro** [comment]
Hi @tonidas,

It's been a while I worked on this but first thing that needs explaining, the need for this `possibleDuplicates` comes from the fact that notion timestamp only resolves to minutes and not seconds or milliseconds like usual timestamps.
In order to trigger on items that were updated in the same minute the trigger happened without retriggering we kept the list of IDs and only retrigger on new ones.
The trade off was that if an item was updated twice in the same minute (one before and one after the trigger), we would miss that update but that was acceptable.

The bug comes from change on an additional commit from @michael-radency that simplified the code but didn't follow the comment that I left above.
https://github.com/n8n-io/n8n/pull/2817/commits/bcd32c56f18f3363772a3c8d0e65302ba40a22df#diff-3011b5287d12171cf0929d60657a12416719c8910d5cf08e79dac19b95f7902cR193

In my original PR I only filtered out duplicates on the items updated at the same minute as the last trigger:
https://github.com/n8n-io/n8n/commit/45399ae1590ee9dea8881f1a9e27778cfe51412d#diff-3011b5287d12171cf0929d60657a12416719c8910d5cf08e79dac19b95f7902cR184

In the added commit the condition on the same minute was removed.

I hope that helps and thanks for researching that bug @tonidas .

**Comment by tonidas** [comment]
Hey @knshiro thank you very much for your response.

This Notion behavior on edited time is just bizarre!

Well, I suggest here a fix on the `possibleDuplicates` and an advisory message at this node's (Notion Trigger) documentation saying that the trigger may ignore events that happened at the same minute.

My use case is just a sync between notion database pages in a specific destination. So, for now I'll stick with a Database Page - Get Many, with the pages edited in the last 10minutes. Processing the same page multiple times for me is not an issue.

Cheers

**Comment by Joffcom** [comment]
I wonder why it was done that way, I will catch Michael on Monday and ask if he remembers.

**Comment by Joffcom** [comment]
We’re taking some time to review and tidy up our issue tracker. It looks like this issue hasn’t had any recent activity for the past 6 months. To keep our backlog focused on current priorities, we’re closing this issue for now.

That said, this doesn’t mean the issue is forgotten! If this is still relevant or you have more information to add, feel free to reopen it or create a new issue and we’d be happy to revisit it.

Thank you for understanding and for contributing to the project!
